Upgrade and add multiple active purges

ananace opened the pull request:
This fixes a bunch of rubocop complaints and adds an env for choosing how many simultaneous active purges is to be used, defaulting to 5.  
The limit avoids overloading either Synapse or the backing Postgres server with actions.

The new output format is a bit more verbose, including a list of all rooms as well as their purge status, along with a progress bar and time elapsed at the bottom.

This is a rather major change, and the use of a progress bar is a bit opinionated, so I'm okay with moving the new output format to behind an environment flag if wanted.

djmaze commented on synapse-purge.rb:
```diff
@@ -3,33 +3,104 @@
+    end
+
+    purges.delete entry
+    purges << entry
```
What are these two lines for?

ananace replied:
Just for visualization purposes, it moves the active entry down to the bottom so it's printed right above the progress bar.
Probably going to redo it though, to just list the currently active purges instead of the status for all rooms.

djmaze commented:
Thanks for the PR! Especially for the style cleanup and client refactorings.

Concerning the multiple active purges and the progressbar, wow, those changes make the code quite hard to read and understand. (We are Rubyists, aren't we? ;)) I really would like to have this optional. Maybe try to make it a separate class, with a better separation of concerns? E.g. as a strategy?

ananace commented:
Have you had any time to look at the added changes?

I split out the printing code into a set of visualizer classes, as that was the part that was causing the most convolution in the code flow.
I also moved the actual purging into a purge worker class to reduce the amount of code in the executing script, and to add a bit of object orienting to it. I thought about doing two worker classes for multiple, and for ==1 active purges, but since 90% of the code would've been identical I decided to just follow DRY and make a single one that I commented a bit extra.

djmaze commented:
@ananace Sorry for the late answer. I will try to have a look tomorrow!

djmaze commented:
Wow, this looks really awesome. And I am so sorry I complete forgot about this PR. (Well, Github does not really help with this..)

Thanks for your work!
